mountpoint-s3/src/fuse/session.rs

```diff
@@ -101,3 +88,249 @@ enum Message {
+    fn clone(&self) -> Self {
+        Self {
+            state: self.state.clone(),
+            workers: self.workers.clone(),
+            max_workers: self.max_workers,
+        }
+    }
+}
+
+impl<FS> Work for Session<FS>
+where
+    FS: Filesystem + Send + Sync + 'static,
+{
+    type Result = io::Result<()>;
+
+    fn run<FB, FA>(&self, mut before: FB, mut after: FA) -> Self::Result
+    where
+        FB: FnMut(),
+        FA: FnMut(),
+    {
+        self.run_and_notify(
+            |req| {
+                // Do not scale threads on bursts of forget messages.
+                if req.is_forget() {
+                    return;
+                }
+                before();
+            },
+            |req| {
+                // Do not scale threads on bursts of forget messages.
+                if req.is_forget() {
+                    return;
+                }
+                after();
+            },
+        )
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn get_thread_id_string() -> String {
+    // SAFETY: this syscall is available since Linux 2.4.11 but glibc didn't
+    // wrap it until very recently.
+    let tid = unsafe { libc::syscall(libc::SYS_gettid) };
+    format!("thread id {tid}")
+}
+
+#[cfg(not(target_os = "linux"))]
+fn get_thread_id_string() -> String {
+    "unknown thread id".to_string()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::sync::{
+        mpsc::{self, Receiver},
+        Condvar, Mutex,
+    };
+    use std::time::Duration;
+    use test_case::test_case;
+
+    use super::*;
+
+    struct TestMessage {
+        _id: usize,
+        mutex: Mutex<bool>,
+        cond: Condvar,
+    }
+
+    impl TestMessage {
+        fn new(_id: usize) -> Self {
+            Self {
+                _id,
+                mutex: Mutex::new(false),
+                cond: Condvar::new(),
+            }
+        }
+
+        fn process(&self) {
+            let mut done = self.mutex.lock().unwrap();
+            while !*done {
+                done = self.cond.wait(done).unwrap();
+            }
+        }
+
+        fn complete(&self) {
+            let mut done = self.mutex.lock().unwrap();
+            *done = true;
+            self.cond.notify_one();
+        }
+    }
+    struct TestWork {
+        receiver: Arc<Mutex<Receiver<Arc<TestMessage>>>>,
+    }
+
+    impl Work for TestWork {
+        type Result = ();
+
+        fn run<FB, FA>(&self, mut before: FB, mut after: FA) -> Self::Result
+        where
+            FB: FnMut(),
+            FA: FnMut(),
+        {
+            while let Ok(message) = {
+                let receiver = self.receiver.lock().unwrap();
+                receiver.recv()
+            } {
+                before();
+                message.process();
+                after();
+            }
+        }
+    }
+
+    #[test_case(10, 10)]
+    #[test_case(10, 30)]
+    #[test_case(30, 10)]
+    fn test_worker_pool(max_worker_threads: usize, concurrent_messages: usize) {
```

Comment:
It took me a long time to parse what this test is doing, so probably deserves a comment. Here's my understanding: it tests that the spawning logic never under-spawns threads, by assigning each worker thread a work item that only completes when a flag is flipped, and then arranges for the flag to flip only once `max_worker_threads` have been spawned. Neat!

I thought maybe you could use `std::sync::Barrier` to make this simpler (set the barrier to `max_worker_threads + 1`), but I guess it doesn't have a `wait_timeout` so you'd block forever if the test broke.

Reply:
Also, I think we should write one other, simpler test: the work item is just incrementing a shared counter. We check that the counter got incremented exactly as many times as we expected, and that the number of spawned threads was no greater than `max_worker_threads`.